### PR TITLE
Add support for Couchbase's role based access

### PR DESCRIPTION
**Pull request by enesacikoglu**
Couchbase role-based access for bucket released after version 5.X.But ,there is no configuration support for Spring Data Couchbase auto configuration after couchbase version 5.X takes username for authentication.

Related problem as linked below;

https://stackoverflow.com/questions/47169384/couchbase-bucket-authentication-error

So, I added username to couchbase properties to solve problem.

### CODE REVIEW

**Comment by pivotal-issuemaster** [comment]
@enesacikoglu Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-projects/spring-boot&pullRequestId=16389)!

[Click here](https://cla.pivotal.io/sync/spring?repositoryId=spring-projects/spring-boot&pullRequestId=16389) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

**Comment by pivotal-issuemaster** [comment]
@enesacikoglu Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-projects/spring-boot&pullRequestId=16389)!

**Comment by snicoll** [comment]
@bsubhashni wondering if you have a bit of time to look at this one.

**Comment by bsubhashni** [comment]
Adding @dnault

**Comment by enesacikoglu** [comment]
Hello,

Waiting for review.

Thanks

**Comment by snicoll** [comment]
@enesacikoglu please be patient, I've already triaged your PR and asked for some feedback from the Couchbase team.

File: spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/couchbase/CouchbaseProperties.java [hunk 1]
```diff
@@ -70,6 +69,16 @@ public Env getEnv() {
...
 
+		/**
+		 * Username of the bucket.
+		 */
+		private String userName = "";
+
+		/**
+		 * RoleBaseAccessEnable for support Couchbase bucket after version 5.0.
+		 */
+		private boolean roleBaseAccessEnabled = false;
```
[the remark this answers is not in the document]
Comment (enesacikoglu) [reply spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/couchbase/CouchbaseProperties.java]: removed roleBaseAccessEnabled properties.

File: spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/couchbase/CouchbaseProperties.java [hunk 2]
```diff
@@ -70,6 +69,16 @@ public Env getEnv() {
 		 */
 		private String password = "";
 
+		/**
+		 * Username of the bucket.
+		 */
+		private String userName = "";
```
[the remark this answers is not in the document]
Comment (enesacikoglu) [reply spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/couchbase/CouchbaseProperties.java]: moved username and password to top level of properties.

File: spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/couchbase/CouchbaseConfiguration.java [hunk 3]
```diff
@@ -57,7 +57,14 @@ public DefaultCouchbaseEnvironment couchbaseEnvironment() {
 	@Bean
 	@Primary
 	public Cluster couchbaseCluster() {
-		return CouchbaseCluster.create(couchbaseEnvironment(), determineBootstrapHosts());
+		CouchbaseCluster couchbaseCluster = CouchbaseCluster
+				.create(couchbaseEnvironment(), determineBootstrapHosts());
+		CouchbaseProperties.Bucket bucket = this.properties.getBucket();
+		if (bucket.isRoleBaseAccessEnabled()) {
+			return couchbaseCluster.authenticate(bucket.getUserName(),
+					bucket.getPassword());
```
[the remark this answers is not in the document]
Comment (enesacikoglu) [reply spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/couchbase/CouchbaseConfiguration.java]: Right thanks :) 
added also here with RBAC control.

**Comment by enesacikoglu** [comment]
> @enesacikoglu Thank you. Support for Role Based Access Control (RBAC) is a much needed feature. Here are some notes.

@dnault thanks for your kindly review and comments.
I try to implement new feature via your suggestions.

Could you review again ?

File: spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/couchbase/CouchbaseConfiguration.java [hunk 4]
```diff
@@ -57,7 +57,14 @@ public DefaultCouchbaseEnvironment couchbaseEnvironment() {
 	@Bean
 	@Primary
 	public Cluster couchbaseCluster() {
-		return CouchbaseCluster.create(couchbaseEnvironment(), determineBootstrapHosts());
+		CouchbaseCluster couchbaseCluster = CouchbaseCluster
+				.create(couchbaseEnvironment(), determineBootstrapHosts());
+		if (this.properties.getUsername() != null
+				&& this.properties.getPassword() != null) {
```
Comment (dnault) [inline spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/couchbase/CouchbaseConfiguration.java]: The `username` and `password` properties are initialized to empty string (`""`). Will they ever be null? A better check might be `.isEmpty()`.

Comment (enesacikoglu) [reply spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/couchbase/CouchbaseConfiguration.java]: Opps how can ı miss 👎 Thanks for your awesome attention

File: spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/couchbase/CouchbaseConfiguration.java [hunk 5]
```diff
@@ -79,6 +86,10 @@ public ClusterInfo couchbaseClusterInfo() {
 	@Bean
 	@Primary
 	public Bucket couchbaseClient() {
+		if (this.properties.getUsername().isEmpty()
+				|| this.properties.getPassword().isEmpty()) {
+			return couchbaseCluster().openBucket(this.properties.getBucket().getName());
```
Comment (dnault) [inline spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/couchbase/CouchbaseConfiguration.java]: This condition should be inverted. If RBAC `username` or `password` are empty we should call the openBucket method that takes a bucket password.

Sorry if I missed this earlier.

Comment (enesacikoglu) [reply spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/couchbase/CouchbaseConfiguration.java]: > This condition should be inverted. If RBAC `username` or `password` are empty we should call the openBucket method that takes a bucket password.
> 
> Sorry if I missed this earlier.

Thanks. I inverted it. 👍

**Comment by snicoll** [comment]
@enesacikoglu thanks for making your first contribution to Spring Boot. This is now merged with a [polish commit](https://github.com/spring-projects/spring-boot/commit/2949561bf6279279089f6cf19a07b195790abb8a).

In particular, I didn't like that username and password are set to blank and that the condition that checks if RBA is enabled wasn't shared. There are now `null` and we check they are both set, as [this article](https://forums.couchbase.com/t/v5-0-new-role-based-authentication-bucket-passwords-etc/14637?u=jkurtz) states that:

> You cannot create a user without a password.

**Comment by snicoll** [comment]
> Support for Role Based Access Control (RBAC) is a much needed feature. 

@dnault in the future, please feel free to reach out to us if there are changes in Couchbase that require a change on our side. Thanks!

**Comment by sadikkuzu** [comment]
Congrats @enesacikoglu 
You're doing the best!

**Pull request closed without merging**